cip/CIP2017-06-18-multiple-graphs.adoc

```diff
@@ -0,0 +1,476 @@
+--
+
+toc::[]
+
+== Motivation
+
+Cypher today is a query language for property graphs that provides access to a single, global, implicit graph in order to extract, transform, and return tabular data that is derived from it.
+While such returned tabular data may include graph entities (such as nodes and relationships), in essence Cypher as a language is not closed under graphs and as a consequence, Cypher queries are not yet (graph) compositional.
+
+However it seems desirable that a language for property graphs should not only enable querying and updating a selected graph but should also support the construction and transformation of multiple graphs, ideally by utilizing a mechanism for incremental query composition.
+
+Furthermore, adding support for working with multiple graphs has recently been identified as a frequently requested feature:
+
+* It would enable the dynamic construction of graph views (e.g. for access control)
+* It allows reasoning over multiple versions of the same graph (e.g. comparing daily snapshots)
+* It may provide a great grouping mechanism for naturally partitioned data (e.g. per-continent graph)
+* It could be used for combining data from disparate data sources in one system (e.g. in federation and data integration scenarios)
+* It fits the paradigm of prominent analytical big-data processing systems (like e.g. Apache Spark)
+* It mirrors mathematical graph theory where working with multiple graphs is standard
+
+== Background and Overview
+
+This proposal needs to modify and extend different aspects of the language consistently.
+
+Therefore background material regarding current semantics is given in line with the different individual sections of this proposal.
+
+== Proposal: Multiple Property Graphs Model
+
+Before proposing changes to the language, it is necessary to first address required changes to the underlying property graph model.
+
+The revised property graph model used in this CIP is called the *multiple property graphs model* and assumes everything specified in the original property graph model extended with the notion of multiple, optionally overlapping graphs outlined in the following.
+
+=== Model definition
+
+This CIP defines an instance of the *multiple property graphs model* to be a set of property graphs that adhere to the definitions given below.
+
+=== (Property) Graph
+
+_Definition_ A *property graph* is a set of labeled nodes and typed relationships both together with their properties (a property is a tuple of a named key and a value).
+Graphs may be updatable, i.e. the set of contained nodes and relationships may change during the lifetime of the graph.
+It is an error to attempt to update a read-only graph.
+
+The same node or relationship may be part of many graphs.
+A relationship may only be part of a graph if it's start node and it's end node are both also part of the same graph.
```

Comment:
Yeah, used to trip me up too, but then I learned that `it's == it is`, so in case you're unsure, just spell it out and it'll become apparent :)